RFC: Specification

dbrgn opened the pull request:
Here's a first draft of the specification. It's still work in progress.

[Rendered](https://github.com/fontobene/fontobene/blob/rfc-specification/SPECIFICATION.md)

dbrgn commented on SPECIFICATION.md:
```diff
@@ -0,0 +1,143 @@
+Every glyph definition starts with the glyph declaration, a unicode codepoint
+in square brackets (e.g. `[1F4A9]` for the pile of poo character). It may
+optionally be followed by a UTF-8 representation of the character for easier
+previewing (e.g. `💩`). However, that preview will be ignored by the parser.
```
Maybe the "comment" after the character codepoint should be the Unicode name (e.g. "PILE OF POO") instead of the representation (e.g. "💩")? That would also work for glyphs not supported by editor / OS font, but it would make the font files larger.

ubruhin replied:
I would prefer the real representation because it keeps the files small... If an editor cannot display some of these characters, that's not really tragic IMO ;)

dbrgn replied:
Ok, let's keep this :)

dbrgn replied:
Another open question: Should we allow metadata for the glyphs? That would allow for things like kerning to be specified and give us some more flexibility in the future.

Since a polyline starts with a digit or with 0, and a reference starts with a capital C, we could simply allow lowercase key-value pairs for metadata anywhere in the glyph definition (but with the convention to put it immediately after the declaration).

dbrgn commented on SPECIFICATION.md:
```diff
@@ -0,0 +1,143 @@
+Because many glyphs are very similar, references may be used as "includes".
+They start with the "@" symbol (U+0040), followed by a codepoint. Here is the
+definition of the glyph for the character "LATIN CAPITAL LETTER A WITH
+CIRCUMFLEX (U+00C2)":
```
Should we use the "@" prefix for includes, or stick with "C", or use something else?

ubruhin replied:
I would stick with "C" for the moment because it's a "simpler" character than "@" and because of the better compatibility to LFF. We could still open an issue some time to discuss about changing that character ;)

dbrgn commented on SPECIFICATION.md:
```diff
@@ -0,0 +1,143 @@
+    # License: GPLv3+
+
+All keys that may be used are specified in the specification. A user MAY use
+non-standard keys, but must prefix them with `X-`, for example `X-Foo: Bar`.
```
Not sure if the `X-` is a good idea. I'd like to have a way to discern standardized from non-standard headers though. Otherwise it could happen that users start defining non-standard headers that become standardized later, resulting in invalid header entries, making the font files non-forwards-compatible.

ubruhin replied:
Why not write non-standardized data as comments, where arbitrary text can be written anyway?

Or if we use a section `[meta]` as described above, we could define a section `[meta_nonstandard]` or something like that, which could be used for non-standardized metadata.

dbrgn commented on SPECIFICATION.md:
```diff
@@ -0,0 +1,143 @@
+The header is parsed line-by-line, starting with a number sign (`#`) as the
+first character on every line. The number sign is followed by a space, a key,
+and a value. Key and value are separated by a colon (`:`). Everything following
+the colon is considered the value. Both keys and values are strings.
```
Should we allow multi-line values? Or should we add a section after the metadata key-value pairs that can be used as free-form text area, e.g. for comments? (See "Notes" [here](https://github.com/LibreCAD/LibreCAD/blob/master/librecad/support/fonts/iso.lff))

ubruhin replied:
Maybe use the `#` character for marking lines as comments, like in programming languages.

For metadata, we could use a similar concept as for the glyphs:
```ini
[meta]
version=1.0
license=GPLv3+
...
```

dbrgn replied:
I like that :) Will update the spec.

ubruhin commented on SPECIFICATION.md:
```diff
@@ -0,0 +1,154 @@
+
+The coordinate system starts at the lower left corner. Positive X coordinates
+point to the right. Positive Y coordinates point to the top. Negative
+coordinates are not allowed.
```
Why are negative values not allowed, resp. where is the baseline of characters like "p", q" and so on?

dbrgn replied:
That's a good point, I'll still have to define that. There are three ways how we could handle alignment and scaling:

- We could use a reference height for characters, e.g. in the range 0 (from the descender) to 100 (at the ascender). The baseline would be e.g. at 20. That would save us from having to parse negative numbers.
- We could allow negative numbers and specify y=0 as the baseline and e.g. 120 as the ascender line. That would also make the height fixed.
- We could allow the negative numbers and specify y=0 as the baseline, without specifying the height of a character. The font would include a `scaling_factor` in the header though, which would be multiplied with the coordinates to achieve a standardized glyph height.

![img](http://www.loveofgraphics.com/wp-content/uploads/2011/10/AnatomyofLetter.png)

ubruhin replied:
At the moment I cannot say which variant I would prefer ;) 

But it seems that the coordinate system is (one of) the most complicated and essential part of our font specification. Maybe it would make sense to document the concept of the coordinate system (incl. images) in a separate markdown file too keep the file format specification more clean. Or what do you think @dbrgn?

Btw, in case of the font is used for PCBs or other fabrication related things, It's very important to have the text size known in a physical length unit (e.g. millimeters). I don't know much about fonts, so I'm not sure how the relation between point, pixel and physical size exactly is. In which unit should the coordinates in the font file be defined? And should the parser provide a way for unit conversion?

dbrgn replied:
> In which unit should the coordinates in the font file be defined?

In none :) I'd just specify a range, e.g. 0-100 for baseline to cap height. Then your software can scale the font as desired.

Having a fixed height range (0-100) requires more work when doing font conversions, but it's much easier to handle than having to deal with a scale factor in the metadata. I think 0-100 for base-cap height is my current favorite. The fonts could go a bit over / below these values if desired, but it's in their responsibility to keep the fonts readable.

ubruhin replied:
> I'd just specify a range, e.g. 0-100 for baseline to cap height. Then your software can scale the font as desired.

Oh, so simple and still powerful :see_no_evil: :smiley: 

But probably we still need to define some more things, because in the LFF header one can find these additional parameters:
```
# LetterSpacing:     3
# WordSpacing:       6.75
# LineSpacingFactor: 1
```

dbrgn replied:
I'm not sure if the font would need to influence spacing. Isn't that something that the "layout engine" should do, depending on glyph width? (The glyph width can easily be determined by doing something like `min(x for (x, y) in pairs)` and `max(x for (x, y) in pairs)`).

ubruhin replied:
I think at least the space character (" ") is a problem, because it's calculated width would be zero (no polylines) which is wrong. LFF seems to define the width of the space character with the attribute "WordSpacing" in the header.

But maybe it would make sense to specify new commands to define manual space in glyphs, something like that:
```
[0020] SPACE
S0,100 # space from 0 to 100 of the x coordinate
```

dbrgn replied:
Yeah, I wonder how we should deal with that. I think the space character is the only one that we need to consider. For all others we don't need any space. It could also be derived from the width of a specific glyph though, e.g. the capital A? Not sure if this is something that is being done by font layout engines or not.

ubruhin commented on SPECIFICATION.md:
```diff
@@ -0,0 +1,154 @@
+be on their own line.
+
+To prevent reference loops and to facilitate single-pass parsers, only backward
+references are allowed. All references glyphs must have been previously defined.
```
To be honest, I would strongly prefer to support forward references because it allows to sort characters by their codepoint, without any space for interpretation. And when updating one character in the font file, no change of the order is required. If forward references are not allowed, the order of characters is not fully defined, and may even be changed if one character starts to inherit from another character --> bad for version control of the font files.

dbrgn replied:
> without any space for interpretation

Could you give an example for "space of interpretation"? What could be mis-interpreted?

> And when updating one character in the font file, no change of the order is required.

A change of order would very rarely be necessary, because unicode already orders characters in a useful way. For example, all latin characters with diacritics are ordered *after* the corresponding latin character.

> bad for version control of the font files

Is this a problem for C files in your experience? C also discourages implicit declarations (it was even invalid in C99).

ubruhin replied:
> Could you give an example for "space of interpretation"? What could be mis-interpreted?

Oh, I think "room for interpretation" would be more correct than "space...", but probably you understood me ;)

So, if no forward references are allowed, the order if glyphs in the font file must be chosen according the inheritance order, which is not 100% canonical (if "A" or "B" comes first is not relevant because they don't inherit from each other, so one can swap them if he likes).

But if the glyphs MUST be ordered by their codepage number, it's like a canonical file format - the creator has no room for choosing some arbitrary order. And I **love** canonical file formats xD

> Is this a problem for C files in your experience?

I don't think one can compare this with C files because in C one can use forward declarations / function prototypes. So the implementation of a function actually *can* call a function whose implementation is unknown, as long as its existence is known...

But let's see the question from another side: Why should we disallow forward references? I think the parser performance and memory usage shouldn't be an issue for the target audience of this font library. And the implementation of forward references is done in a few minutes (see LFF parser). So what's the reason to disallow them?

dbrgn replied:
> So what's the reason to disallow them?

The parser is simpler to write, simpler to understand and simpler to maintain.

@rnestler what do you think?

dbrgn commented on SPECIFICATION.md:
```diff
@@ -0,0 +1,177 @@
+    1.5,11.5;3,13;4.5,11.5
+
+An implementation would simply replace the "C0041" part with the expanded
+polylines from the referenced glyph.
```
Should we add a possibility to define meta-glyphs that are only used for includes, but are not valid glyphs themselves? The circumflex is a bad example because it actually is a valid character, but there could be others.

Also, should we make it possible to translate and/or scale includes? E.g. with `C0041 offset=0.5,1.5 scale=1.2`

ubruhin commented:
@dbrgn Great work so far! Maybe we should already merge this PR soon, and then start discussing the open questions in separate issues?

But first, we probably need to add some kind of note to make clear that this specification is not yet considered as stable (the version number "1.0" in the title of the file looks like it is a stable release).

ubruhin commented on SPECIFICATION.md:
```diff
@@ -0,0 +1,177 @@
+| --- | --- | --- | --- |
+| name | The name of this font. | 1 | `LibrePCB Font` |
+| id | The identifier of this font. MUST only contain lowercase letters and minus characters. MUST start and end with a lowercase letter. | 1 | `librepcb` |
+| version | The version of this font. SHOULD follow semantic versioning. | 0-1 | `0.4.1` |
```
Shouldn't the version number be mandatory? I think the version is quite important, and if it's not used one can just write "1.0" into this field ;)

dbrgn commented:
> @dbrgn Great work so far! Maybe we should already merge this PR soon, and then start discussing the open questions in separate issues?

Ok, we can do that!

dbrgn commented:
> the version number "1.0" in the title of the file looks like it is a stable release

I changed that to 1.0-WIP and moved all discussions to separate issues.
